protocol/03-options.md
# Options

Options are common objects for use across different page types. Some options may be required for a given page to be parsed or accepted into the database.
Option Kind and Length are always specified for backwards / cross compatibility to allow parsers to skip unrecognised options.

**Structure:**

```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|          Option Kind          |         Option Length         |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
/                                                               /
/                          Option Data                          /
/             Variable Length Data (32-bit aligned)             /
/                                                               /
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

**Kinds:**

| Name                      | ID   | Description                | Multiple? | Size     |
| ------------------------- | ---- | -------------------------- | --------- | -------- |
| [PubKey](### PubKey)    | 0x00 | Public Key                 | No        | 32-byte  |
| [DatabaseId]()            | 0x01 | Database (Peer or Page) ID | No        | 32-byte  |
| [RequestId]()             | 0x02 | Message Request ID         | No        | 8-byte   |
| [Kind]()                  | 0x03 | Arbitrary Service Kind     | No        | Variable |
| [Name]()                  | 0x04 | Arbitrary Service Name     | No        | Variable |
| [V4Addr]()                | 0x05 | IPv4 address and port      | Yes       | 10 byte  |
| [V6Addr]()                | 0x06 | IPv6 address and port      | Yes       | 18 byte  |
| [Issued]()                | 0x07 | Issued timestamp           | No        | 8 byte   |
| [Expiry]()                | 0x08 | Expirty timestamp          | No        | 8 byte   |
| [Metadata]()              | 0x09 | Metadata Key-Value pair    | Yes       | Variable |




## PubKey

The cryptographic public key associated with a service or peer.

```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|              0x00             |               32              |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                                                               |
/                           Public Key                          /
/                    32-byte ECDSA Public Key                   /
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

## DatabaseId

The Database ID of a service, peer or page

```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|              0x01             |               32              |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                                                               |
/                            Peer ID                            /
/               32-byte Protocol Defined ID Length              /
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

## Request ID

A Request ID used for pairing requests and responses

```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|              0x02             |               16              |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                                                               |
|                           Request ID                          |
|                   16-byte Request ID Length                   |
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

## Service Kind

Arbitrary service kind, a string that identifies a type of service, for example: `mqtt` for an MQTT broker.
These are not required to be globally unique or consistent across different users (unless interoperability is desired).

Note that for interoperability or more complex services with data, page kinds should be used in favour of this field.

```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|              0x03             |          Kind Length          |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                                                               |
/                   Kind (utf8 encoded string)                  /
/                        Variable Length                        /
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

## Service Name

Arbitrary service name, a user defined string to identify a service, for example: `home-web` for a home web server.
These are not required to be globally unique or consistent across different users.

```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|              0x04             |          Name Length          |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                                                               |
/                   Name (utf8 encoded string)                  /
/                        Variable Length                        /
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```


## IPv4 Address

An IPv4 Address and Port for connecting to a service or peer.

```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|              0x05             |               10              |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                          IPv4 Address                         |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|              Port             |            Reserved           |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

## IPv6 Address

An IPv6 Address and Port for connecting to a service or peer.

```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|              0x06             |               10              |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                                                               |
|                                                               |
|                                                               |
|                          IPv6 Address                         |
|                                                               |
|                                                               |
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|              Port             |            Reserved           |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

## Issued

A timestamp at which the page was issued as a 64-bit little-endian uint representing milliseconds from the unix epoc.

```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|              0x07             |               8               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                            Issued                             |
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

## Expiry

A timestamp at which the page should expire as a 64-bit little-endian uint representing milliseconds from the unix epoc.

```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|              0x08             |               8               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                            Expiry                             |
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

## Metadata

Arbitrary Key:Value pairs as UTF-8 strings, separated using the `|` character, this character may not be used in the Key or Value terms.

```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|              0x09             |           Variable            |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
/                                                               /
/                             String                            /
/                                                               /
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```



src/protocol/options.rs
use std::io::{Cursor, Write, Read};
use std::ops::Add;
use std::str;
use std::time::{Duration, SystemTime};
use std::net::{SocketAddr, SocketAddrV4, SocketAddrV6};

use byteorder::{ByteOrder, NetworkEndian, ReadBytesExt, WriteBytesExt};

use crate::types::{Id, ID_LEN, PublicKey, PUBLIC_KEY_LEN, RequestId, REQUEST_ID_LEN};
use crate::protocol::{Encode, Parse};

/// D-IoT Optional fields
#[derive(Debug, Clone, PartialEq)]
pub enum Options {
    None,
    PubKey(PubKey),
    PeerId(PeerId),
    RequestId(ReqId),
    Kind(Kind),
    Name(Name),
    IPv4(SocketAddrV4),
    IPv6(SocketAddrV6),
    Issued(Issued),
    Expiry(Expiry),
    Metadata(Metadata),
}

#[derive(Debug, Clone, PartialEq)]
pub enum OptionsError {
    IO(std::io::ErrorKind),
    InvalidMetadata,
    InvalidOptionLength,
    Unimplemented,
}

impl From<std::io::Error> for OptionsError {
    fn from(e: std::io::Error) -> OptionsError {
        OptionsError::IO(e.kind())
    }
}

/// D-IoT Option types
mod option_kinds {
    pub const PUBKEY:       u16 = 0x00; // Public Key
    pub const PEER_ID:      u16 = 0x01; // ID of Peer responsible for secondary page
    pub const REQUEST_ID:   u16 = 0x02;
    pub const KIND:         u16 = 0x03; // Service KIND in utf-8
    pub const NAME:         u16 = 0x04; // Service NAME in utf-8
    pub const ADDR_IPV4:    u16 = 0x05; // IPv4 service address
    pub const ADDR_IPV6:    u16 = 0x06; // IPv6 service address
    pub const ISSUED:       u16 = 0x07; // ISSUED option defines object creation time
    pub const EXPIRY:       u16 = 0x08; // EXPIRY option defines object expiry time
    pub const META:         u16 = 0x09; // META option supports generic metadata key:value pairs
}

const OPTION_HEADER_LEN: usize = 4;

impl Options {
    /// Parse a bounded list of options into a vector
    pub fn parse_vec(data: &[u8]) -> Result<(Vec<Options>, usize), OptionsError> {
        let mut options = Vec::new();
        let mut rem = &data[..];
        let mut i = 0;

        while rem.len() >= OPTION_HEADER_LEN {
            let (o, n) = Options::parse(&rem)?;
            i += n;

            options.push(o);
            rem = &data[i..];
        }

        Ok((options, i))
    }

    /// Encode a vector of options
    pub fn encode_vec(options: &Vec<Options>, data: &mut [u8]) -> Result<usize, OptionsError> {
        let mut i = 0;

        for o in options.iter() {
            i += o.encode(&mut data[i..])?;
        }

        Ok(i)
    }
}

impl Options {
    // Helper to generate name metadata
    pub fn name(value: &str) -> Options {
        Options::Name(Name::new(value.clone()))
    }

    pub fn kind(value: &str) -> Options {
        Options::Kind(Kind::new(value.clone()))
    }

    pub fn meta(key: &str, value: &str) -> Options {
        Options::Metadata(Metadata::new(key.clone(), value.clone()))
    }

    pub fn issued(now: SystemTime) -> Options {
        Options::Issued(Issued::new(now))
    }

    pub fn expiry(when: SystemTime) -> Options {
        Options::Expiry(Expiry::new(when))
    }

    pub fn peer_id(id: Id) -> Options {
        Options::PeerId(PeerId::new(id))
    }

    pub fn public_key(public_key: PublicKey) -> Options {
        Options::PubKey(PubKey::new(public_key))
    }

    pub fn address<T>(address: T) -> Options 
    where T: Into<SocketAddr>
    {
        match address.into() {
            SocketAddr::V4(v4) => Options::IPv4(v4),
            SocketAddr::V6(v6) => Options::IPv6(v6),
        }
    }
}

/// Parse parses a control option from the given scope
impl Parse for Options {
    type Output = Options;
    type Error = OptionsError;

    fn parse<'a>(data: &'a [u8]) -> Result<(Self::Output, usize), Self::Error> {

        if data.len() < OPTION_HEADER_LEN {
            return Err(OptionsError::InvalidOptionLength);
        }

        let option_kind = NetworkEndian::read_u16(&data[0..2]);
        let option_len = NetworkEndian::read_u16(&data[2..4]) as usize;

        let d = &data[OPTION_HEADER_LEN..OPTION_HEADER_LEN + option_len];

        match option_kind {
            option_kinds::PUBKEY => {
                let (opt, n) = PubKey::parse(d)?;
                Ok((Options::PubKey(opt), n + OPTION_HEADER_LEN))
            },
            option_kinds::PEER_ID => {
                let (opt, n) = PeerId::parse(d)?;
                Ok((Options::PeerId(opt), n + OPTION_HEADER_LEN))
            },
            option_kinds::KIND => {
                let (opt, n) = Kind::parse(d)?;
                Ok((Options::Kind(opt), n + OPTION_HEADER_LEN))
            },
            option_kinds::NAME => {
                let (opt, n) = Name::parse(d)?;
                Ok((Options::Name(opt), n + OPTION_HEADER_LEN))
            },
            option_kinds::ADDR_IPV4 => {
                let (opt, n) = SocketAddrV4::parse(d)?;
                Ok((Options::IPv4(opt), n + OPTION_HEADER_LEN))
            },
            option_kinds::ADDR_IPV6 => {
                let (opt, n) = SocketAddrV6::parse(d)?;
                Ok((Options::IPv6(opt), n + OPTION_HEADER_LEN))
            },
            option_kinds::META => {
                let (opt, n) = Metadata::parse(d)?;
                Ok((Options::Metadata(opt), n + OPTION_HEADER_LEN))
            },
            option_kinds::ISSUED => {
                let (opt, n) = Issued::parse(d)?;
                Ok((Options::Issued(opt), n + OPTION_HEADER_LEN))
            },
            option_kinds::EXPIRY => {
                let (opt, n) = Expiry::parse(d)?;
                Ok((Options::Expiry(opt), n + OPTION_HEADER_LEN))
            },
            _ => {
                // Unrecognised option types (and None) are skipped
                Ok((Options::None, OPTION_HEADER_LEN + option_len))
            }
        }
    }
}

impl Encode for Options {
    type Error = OptionsError;

    fn encode(&self, data: &mut [u8]) -> Result<usize, Self::Error> {
        match *self {
            Options::PubKey(ref o) => {
                Ok(o.encode(data)?)
            },
            Options::PeerId(ref o) => {
                Ok(o.encode(data)?)
            },
            Options::Kind(ref o) => {
                Ok(o.encode(data)?)
            },
            Options::Name(ref o) => {
                Ok(o.encode(data)?)
            },
            Options::IPv4(ref o) => {
                Ok(o.encode(data)?)
            },
            Options::IPv6(ref o) => {
                Ok(o.encode(data)?)
            },
            Options::Metadata(ref o) => {
                Ok(o.encode(data)?)
            },
            Options::Issued(ref o) => {
                Ok(o.encode(data)?)
            },
            Options::Expiry(ref o) => {
                Ok(o.encode(data)?)
            },
            _ => {
                Err(OptionsError::Unimplemented)
            }
        }
    }
}


#[derive(Debug, Clone, PartialEq)]
pub struct PubKey {
    pub public_key: PublicKey,
}

impl PubKey {
    pub fn new(public_key: PublicKey) -> PubKey {
        PubKey { public_key }
    }
}

impl Parse for PubKey {
    type Output = PubKey;
    type Error = OptionsError;

    fn parse(data: &[u8]) -> Result<(Self::Output, usize), Self::Error> {
        let mut public_key = [0u8; PUBLIC_KEY_LEN];
        public_key.copy_from_slice(&data[..PUBLIC_KEY_LEN]);

        Ok((PubKey { public_key }, PUBLIC_KEY_LEN))
    }
}

impl Encode for PubKey {
    type Error = OptionsError;

    fn encode(&self, data: &mut [u8]) -> Result<usize, Self::Error> {
        let mut w = Cursor::new(data);
        
        w.write_u16::<NetworkEndian>(option_kinds::PUBKEY)?;
        w.write_u16::<NetworkEndian>(PUBLIC_KEY_LEN as u16)?;
        w.write(&self.public_key)?;

        Ok(w.position() as usize)
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct PeerId {
    peer_id: Id,
}

impl PeerId {
    pub fn new(peer_id: Id) -> PeerId {
        PeerId { peer_id }
    }
}

impl Parse for PeerId {
    type Output = PeerId;
    type Error = OptionsError;

    fn parse(data: &[u8]) -> Result<(Self::Output, usize), Self::Error> {
        let mut peer_id = [0u8; ID_LEN];
        peer_id.copy_from_slice(&data[..ID_LEN]);
        Ok((PeerId { peer_id }, ID_LEN))
    }
}

impl Encode for PeerId {
    type Error = OptionsError;

    fn encode(&self, data: &mut [u8]) -> Result<usize, Self::Error> {
        let mut w = Cursor::new(data);
        
        w.write_u16::<NetworkEndian>(option_kinds::PEER_ID)?;
        w.write_u16::<NetworkEndian>(ID_LEN as u16)?;
        w.write(&self.peer_id)?;

        Ok(w.position() as usize)
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct ReqId {
    request_id: RequestId,
}

impl ReqId {
    pub fn new(request_id: RequestId) -> ReqId {
        ReqId { request_id }
    }
}

impl Parse for ReqId {
    type Output = ReqId;
    type Error = OptionsError;

    fn parse(data: &[u8]) -> Result<(Self::Output, usize), Self::Error> {
        let mut request_id = [0u8; REQUEST_ID_LEN];
        request_id.copy_from_slice(&data[..REQUEST_ID_LEN]);
        Ok((ReqId { request_id }, REQUEST_ID_LEN))
    }
}

impl Encode for ReqId {
    type Error = OptionsError;

    fn encode(&self, data: &mut [u8]) -> Result<usize, Self::Error> {
        let mut w = Cursor::new(data);
        
        w.write_u16::<NetworkEndian>(option_kinds::REQUEST_ID)?;
        w.write_u16::<NetworkEndian>(REQUEST_ID_LEN as u16)?;
        w.write(&self.request_id)?;

        Ok(w.position() as usize)
    }
}


#[derive(Debug, Clone, PartialEq)]
pub struct Kind {
    value: String,
}

impl Kind {
    pub fn new(value: &str) -> Kind {
        Kind {
            value: value.to_owned(),
        }
    }
}

impl Parse for Kind {
    type Output = Kind;
    type Error = OptionsError;

    fn parse(data: &[u8]) -> Result<(Self::Output, usize), Self::Error> {

        let value = str::from_utf8(&data).unwrap().to_owned();

        Ok((Kind { value }, data.len()))
    }
}

impl Encode for Kind {
    type Error = OptionsError;

    fn encode(&self, data: &mut [u8]) -> Result<usize, Self::Error> {
        let mut w = Cursor::new(data);

        w.write_u16::<NetworkEndian>(option_kinds::KIND)?;
        w.write_u16::<NetworkEndian>(self.value.len() as u16)?;
        w.write(self.value.as_bytes())?;

        Ok(w.position() as usize)
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Name {
    value: String,
}

impl Name {
    pub fn new(value: &str) -> Name {
        Name {
            value: value.to_owned(),
        }
    }
}

impl Parse for Name {
    type Output = Name;
    type Error = OptionsError;

    fn parse(data: &[u8]) -> Result<(Self::Output, usize), Self::Error> {

        let value = str::from_utf8(&data).unwrap().to_owned();

        Ok((Name { value }, data.len()))
    }
}

impl Encode for Name {
    type Error = OptionsError;

    fn encode(&self, data: &mut [u8]) -> Result<usize, Self::Error> {
        let mut w = Cursor::new(data);

        w.write_u16::<NetworkEndian>(option_kinds::NAME)?;
        w.write_u16::<NetworkEndian>(self.value.len() as u16)?;
        w.write(self.value.as_bytes())?;

        Ok(w.position() as usize)
    }
}

impl Parse for SocketAddrV4 {
    type Output = SocketAddrV4;
    type Error = OptionsError;

    fn parse(data: &[u8]) -> Result<(Self::Output, usize), Self::Error> {
        let mut r = Cursor::new(data);
        
        let mut ip = [0u8; 4];
        r.read(&mut ip)?;
        let port = r.read_u16::<NetworkEndian>()?;

        Ok((SocketAddrV4::new(ip.into(), port), data.len()))
    }
}

impl Encode for SocketAddrV4 {
    type Error = OptionsError;

    fn encode(&self, data: &mut [u8]) -> Result<usize, Self::Error> {
        let mut w = Cursor::new(data);

        w.write_u16::<NetworkEndian>(option_kinds::ADDR_IPV4)?;
        w.write_u16::<NetworkEndian>(6)?;
        w.write(&self.ip().octets())?;
        w.write_u16::<NetworkEndian>(self.port())?;

        Ok(w.position() as usize)
    }
}

impl Parse for SocketAddrV6 {
    type Output = SocketAddrV6;
    type Error = OptionsError;

    fn parse(data: &[u8]) -> Result<(Self::Output, usize), Self::Error> {
        let mut r = Cursor::new(data);
        
        let mut ip = [0u8; 16];
        r.read(&mut ip)?;
        let port = r.read_u16::<NetworkEndian>()?;

        Ok((SocketAddrV6::new(ip.into(), port, 0, 0), data.len()))
    }
}

impl Encode for SocketAddrV6 {
    type Error = OptionsError;

    fn encode(&self, data: &mut [u8]) -> Result<usize, Self::Error> {
        let mut w = Cursor::new(data);

        w.write_u16::<NetworkEndian>(option_kinds::ADDR_IPV6)?;
        w.write_u16::<NetworkEndian>(18)?;
        w.write(&self.ip().octets())?;
        w.write_u16::<NetworkEndian>(self.port())?;

        Ok(w.position() as usize)
    }
}


#[derive(Debug, Clone, PartialEq)]
pub struct Metadata {
    key: String,
    value: String,
}

impl Metadata {
    pub fn new(key: &str, value: &str) -> Metadata {
        Metadata {
            key: key.to_owned(),
            value: value.to_owned(),
        }
    }
}

impl Parse for Metadata {
    type Output = Metadata;
    type Error = OptionsError;

    fn parse(data: &[u8]) -> Result<(Self::Output, usize), Self::Error> {

        let kv = str::from_utf8(&data).unwrap().to_owned();
        let split: Vec<_> = kv.split("|").collect();
        if split.len() != 2 {
            return Err(OptionsError::InvalidMetadata)
        }

        Ok((Metadata { key: split[0].to_owned(), value: split[1].to_owned() }, data.len()))
    }
}

impl Encode for Metadata {
    type Error = OptionsError;

    fn encode(&self, data: &mut [u8]) -> Result<usize, Self::Error> {
        let mut w = Cursor::new(data);
        
        let data = format!("{}|{}", self.key, self.value);

        w.write_u16::<NetworkEndian>(option_kinds::META)?;
        w.write_u16::<NetworkEndian>(data.len() as u16)?;
        w.write(data.as_bytes())?;

        Ok(w.position() as usize)
    }
}

impl From<Metadata> for Options {
    fn from(m: Metadata) -> Options {
        Options::Metadata(m)
    }
}

#[derive(Debug, Clone)]
pub struct Issued {
    when: SystemTime,
}

impl Issued {
    pub fn new(when: SystemTime) -> Issued {
        Issued { when }
    }

    pub fn as_secs(&self) -> u64 {
        let when = self.when.duration_since(SystemTime::UNIX_EPOCH).unwrap();
        when.as_secs()
    }
}

impl PartialEq for Issued {
    fn eq(&self, other: &Issued) -> bool {
        self.as_secs() == other.as_secs()
    }
}

impl Parse for Issued {
    type Output = Issued;
    type Error = OptionsError;
    fn parse<'a>(data: &'a [u8]) -> Result<(Self::Output, usize), Self::Error> {
        let mut r = Cursor::new(data);

        let raw = r.read_u64::<NetworkEndian>()?;
        let duration = Duration::from_secs(raw);
        let when = SystemTime::UNIX_EPOCH.add(duration);

        Ok((Issued { when }, r.position() as usize))
    }
}

impl Encode for Issued {
    type Error = OptionsError;
    fn encode(&self, data: &mut [u8]) -> Result<usize, Self::Error> {
        let mut w = Cursor::new(data);

        w.write_u16::<NetworkEndian>(option_kinds::ISSUED)?;
        w.write_u16::<NetworkEndian>(8)?;

        w.write_u64::<NetworkEndian>(self.as_secs())?;

        Ok(w.position() as usize)
    }
}

#[derive(Debug, Clone)]
pub struct Expiry {
    when: SystemTime,
}

impl Expiry {
    pub fn new(when: SystemTime) -> Expiry {
        Expiry { when }
    }

    pub fn as_secs(&self) -> u64 {
        let when = self.when.duration_since(SystemTime::UNIX_EPOCH).unwrap();
        when.as_secs()
    }
}

impl PartialEq for Expiry {
    fn eq(&self, other: &Expiry) -> bool {
        self.as_secs() == other.as_secs()
    }
}

impl Parse for Expiry {
    type Output = Expiry;
    type Error = OptionsError;
    fn parse<'a>(data: &'a [u8]) -> Result<(Self::Output, usize), Self::Error> {
        let mut r = Cursor::new(data);

        let raw = r.read_u64::<NetworkEndian>()?;
        let duration = Duration::from_secs(raw);
        let when = SystemTime::UNIX_EPOCH.add(duration);

        Ok((Expiry { when }, r.position() as usize))
    }
}

impl Encode for Expiry {
    type Error = OptionsError;
    fn encode(&self, data: &mut [u8]) -> Result<usize, Self::Error> {
        let mut w = Cursor::new(data);

        w.write_u16::<NetworkEndian>(option_kinds::EXPIRY)?;
        w.write_u16::<NetworkEndian>(8)?;

        w.write_u64::<NetworkEndian>(self.as_secs())?;

        Ok(w.position() as usize)
    }
}

#[cfg(test)]
mod tests {

    use super::*;

    use std::net::{Ipv4Addr, Ipv6Addr};

    #[test]
    fn encode_decode_option_types() {
        let tests = [
            Options::PubKey(PubKey::new([1u8; PUBLIC_KEY_LEN])),
            Options::PeerId(PeerId::new([2u8; ID_LEN])),
            Options::Kind(Kind::new("test-kind")),
            Options::Name(Name::new("test-name")),
            Options::IPv4(SocketAddrV4::new(Ipv4Addr::new(127, 0, 0, 1), 8080)),
            Options::IPv6(SocketAddrV6::new(Ipv6Addr::new(0x2001, 0xdb8, 0, 0, 0, 0, 0, 1), 8080, 0, 0)),
            Options::Metadata(Metadata::new("test-key", "test-value")),
            Options::Issued(Issued::new(SystemTime::now())),
            Options::Expiry(Expiry::new(SystemTime::now())),
        ];

        for o in tests.iter() {
            let mut data = vec![0u8; 1024];
            let n1 = o.encode(&mut data).expect(&format!("Error encoding {:?}", o));

            let (decoded, n2) = Options::parse(&data).expect(&format!("Error decoding {:?}", o));

            assert_eq!(n1, n2, "Mismatch between encoded and decoded lengths for object: {:?}", o);
            assert_eq!(o, &decoded, "Mismatch between original and decode objects");
        }
    }

    #[test]
    fn encode_decode_option_list() {
        let tests = vec![
            Options::PubKey(PubKey::new([1u8; PUBLIC_KEY_LEN])),
            Options::PeerId(PeerId::new([2u8; ID_LEN])),
            Options::Kind(Kind::new("test-kind")),
            Options::Name(Name::new("test-name")),
            Options::IPv4(SocketAddrV4::new(Ipv4Addr::new(127, 0, 0, 1), 8080)),
            Options::IPv6(SocketAddrV6::new(Ipv6Addr::new(0x2001, 0xdb8, 0, 0, 0, 0, 0, 1), 8080, 0, 0)),
            Options::Metadata(Metadata::new("test-key", "test-value")),
            Options::Issued(Issued::new(SystemTime::now())),
            Options::Expiry(Expiry::new(SystemTime::now())),
        ];

        let mut data = vec![0u8; 1024];
        let n1 = Options::encode_vec(&tests, &mut data).expect("Error encoding options vector");

        let encoded = &data[0..n1];
        let (decoded, n2) = Options::parse_vec(encoded).expect("Error decoding options vector");
            
        assert_eq!(n1, n2, "Mismatch between encoded and decoded length");
        assert_eq!(&tests, &decoded, "Mismatch between original and decode vectors");
    }
}
